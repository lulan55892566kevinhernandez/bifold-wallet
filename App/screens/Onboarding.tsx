import { useFocusEffect } from '@react-navigation/native'
import React, { Ref, useCallback, useRef, useState } from 'react'
import { Animated, BackHandler, Dimensions, FlatList, StyleSheet, Text, TouchableHighlight, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { SvgProps } from 'react-native-svg'

import { Pagination } from '../components/shared/Pagination'
import { Colors } from '../globalStyles'

const { width } = Dimensions.get('window')

interface IOnboardingStyleSheet {
  container: Record<string, any>
  controlsContainer: Record<string, any>
  headerText: Record<string, any>
  bodyText: Record<string, any>
  primaryButton: Record<string, any>
  primaryButtonText: Record<string, any>
}

interface IOnboardingCarouselData {
  image: React.FC<SvgProps>
  text: string
}

interface IOnboardingProps {
  title: string
  pages: Array<IOnboardingCarouselData>
  onOnboardingDismissed: () => Promise<void>
  style?: IOnboardingStyleSheet
}

const defaultStyle: IOnboardingStyleSheet = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
  controlsContainer: {
    alignItems: 'center',
    marginTop: 16,
    marginBottom: 32,
  },
  headerText: {
    fontWeight: 'bold',
    fontSize: 32,
    color: Colors.textColor,
  },
  bodyText: {
    fontWeight: 'normal',
    fontSize: 32,
    textAlign: 'center',
    margin: 16,
    color: Colors.textColor,
  },
  primaryButton: {
    padding: 16,
    borderRadius: 4,
    backgroundColor: Colors.mainColor,
  },
  primaryButtonText: {
    fontSize: 18,
    color: Colors.textColor,
  },
})

const Onboarding: React.FC<IOnboardingProps> = ({ title, pages, onOnboardingDismissed, style }) => {
  const [activeIndex, setActiveIndex] = useState(0)
  const flatList: Ref<FlatList> = useRef(null)
  const scrollX = useRef(new Animated.Value(0)).current
  const myStyle = style ? style : defaultStyle

  const onViewableItemsChangedRef = useRef(({ viewableItems }: any) => {
    setActiveIndex(viewableItems[0].index)
  })

  const viewabilityConfigRef = useRef({
    viewAreaCoveragePercentThreshold: 60,
  })

  const onScroll = Animated.event([{ nativeEvent: { contentOffset: { x: scrollX } } }], {
    useNativeDriver: false,
  })

  const next = () => {
    if (activeIndex + 1 < pages.length) {
      flatList?.current?.scrollToIndex({
        index: activeIndex + 1,
        animated: true,
      })
    }
  }

  const previous = () => {
    if (activeIndex !== 0) {
      flatList?.current?.scrollToIndex({
        index: activeIndex - 1,
        animated: true,
      })
    }
  }

  const renderItem = useCallback(
    ({ item, index }: { item: { image: React.FC<SvgProps>; text: string }; index: number }) => (
      <View key={index} style={[{ width }, { alignItems: 'center' }]}>
        {item.image({
          fill: Colors.textColor,
          height: 180,
          width: 180,
        })}
        <Text style={myStyle.bodyText}>{item.text}</Text>
      </View>
    ),
    []
  )

  useFocusEffect(
    useCallback(() => {
      const onBackPress = () => {
        BackHandler.exitApp()
        return true
      }
      BackHandler.addEventListener('hardwareBackPress', onBackPress)
      return () => BackHandler.removeEventListener('hardwareBackPress', onBackPress)
    }, [])
  )

  return (
    <SafeAreaView style={myStyle.container}>
      <Text style={myStyle.headerText}>{title}</Text>
      <FlatList
        ref={flatList}
        horizontal
        pagingEnabled
        showsHorizontalScrollIndicator={false}
        style={[{ width }]}
        data={pages}
        renderItem={renderItem}
        viewabilityConfig={viewabilityConfigRef.current}
        onViewableItemsChanged={onViewableItemsChangedRef.current}
        onScroll={onScroll}
        scrollEventThrottle={16}
      />
      <Pagination data={pages} scrollX={scrollX} next={next} previous={previous} />
      <View style={myStyle.controlsContainer}>
        <TouchableHighlight
          style={myStyle.primaryButton}
          underlayColor={Colors.activeMain}
          onPress={onOnboardingDismissed}
        >
          <Text style={myStyle.primaryButtonText}>Get Started!</Text>
        </TouchableHighlight>
      </View>
    </SafeAreaView>
  )
}

export default Onboarding
