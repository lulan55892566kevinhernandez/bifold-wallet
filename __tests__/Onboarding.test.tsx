import React from 'react'
import { SvgProps } from 'react-native-svg'
import renderer from 'react-test-renderer'

import Onboarding from '../App/screens/Onboarding'

const title = 'My Wallet'
const pages: { image: React.FC<SvgProps>; text: string }[] = []
const markTutorialFin = async () => {
  return
}

jest.mock('@react-navigation/native', () => {
  return {
    useFocusEffect: jest.fn(),
  }
})

describe('Onboarding', () => {
  // beforeEach(() => {
  //   // Alternatively, set "clearMocks" in your Jest config to "true"
  //   mockedDispatch.mockClear()
  // })

  it('Renders correctly', () => {
    const tree = renderer
      .create(<Onboarding title={title} pages={pages} onOnboardingDismissed={markTutorialFin} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
