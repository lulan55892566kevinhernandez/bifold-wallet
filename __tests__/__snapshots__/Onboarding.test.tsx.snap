// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding Renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontSize": 32,
        "fontWeight": "bold",
      }
    }
  >
    cake
  </Text>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "width": 750,
        },
      ]
    }
    viewabilityConfig={
      Object {
        "viewAreaCoveragePercentThreshold": 60,
      }
    }
    viewabilityConfigCallbackPairs={
      Array [
        Object {
          "onViewableItemsChanged": [Function],
          "viewabilityConfig": Object {
            "viewAreaCoveragePercentThreshold": 60,
          },
        },
      ]
    }
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "paddingBottom": 16,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <
        fill="#35823f"
        height={24}
        width={48}
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "center",
            "bottom": 20,
            "flexDirection": "row",
            "position": "absolute",
          },
          Object {
            "position": "relative",
            "top": 0,
          },
        ]
      }
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <
        fill="#35823f"
        height={24}
        style={
          Object {
            "transform": Array [
              Object {
                "rotate": "180deg",
              },
            ],
          }
        }
        width={48}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginBottom": 32,
        "marginTop": 16,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#35823f",
          "borderRadius": 4,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
          }
        }
      >
        Get Started!
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
